### ISSUE TITLE
PROPOSAL: PolicyRegistry

### ISSUE THREAD

**Issue by reisenberger** [App-vNext/Polly#226]
A `PolicyRegistry` with `Dictionary<string, Policy>`-like semantics, for storing configured policies and retrieving them for use elsewhere.

Useful for those who want to define policies centrally on start-up, then provide them to components from a registry, by eg dependency-injection or ambient context pattern.

_Thanks to @ankitbko for discussion which influenced the proposal!_

### Hide underlying implementation

`PolicyRegistry : Dictionary<string, Policy>` while simple would tie us into an implementation.  Better a shallow wrapper, something [like this](https://github.com/BrighterCommand/Paramore.Brighter/blob/master/paramore.brighter.commandprocessor/PolicyRegistry.cs), allowing changing underlying implementation.

### `IDictionary<TKey, TValue>` semantics, or custom?

Should we have custom-semantics (as eg [this shallow wrapper](https://github.com/BrighterCommand/Paramore.Brighter/blob/master/paramore.brighter.commandprocessor/PolicyRegistry.cs))?  Or should we implement `IDictionary<TKey, TValue>`, for entirely standard semantics?  

_Suggest_: standard semantics?

#### `Dictionary<TKey, TValue>`

If we accept the majority Use Case is likely to be establish policy registry at application startup on a single-thread, read on multiple threads thereafter (ie no concurrent writes), **`Dictionary<TKey, TValue>`** would be adequate; possibly also (fractionally) most read-performant.  

If this, **document** that the default implementation is optimised for this majority Use Case but not safe for concurrent writes.  

We could also provide segregated `IReadOnlyPolicyRegistry` and `IWritablePolicyRegistry` interfaces; recommend pass around only an `IReadOnlyPolicyRegistry` after populating the registry.

#### `ImmutableDictionary<TKey, TValue>`

**`ImmutableDictionary<TKey, TValue>`**: the best fit for write-extensively-at-startup, read-mostly thereafter (but still safe for occasional writes).  Requires tho Polly taking a dependency on `System.Collections.Immutable` (violates no dependencies principle).  Offer this implementation (if demand) as a separate nuget package?

#### `ConcurrentDictionary<TKey, TValue>` 

**`ConcurrentDictionary<TKey, TValue>`**: eliminates all concurrency risks but requires changing our minimum .NET Standard support from .NET Standard 1.0, to .NET Standard 1.1 (losing .NET Standard 1.0-only users: Windows Phone Silverlight 8?).  Fractionally lower read-performance than `Dictionary<TKey, TValue>`? (perhaps not relevant for likely size of dictionary)  

_Community views?_

Option: a `Dictionary<TKey, TValue>`-based implementation in the main Polly package, but offer `Polly.Registry.Immutable` and `Polly.Registry.Concurrent` (if demand) as separate nuget packages.  

 
### `IPolicyRegistry` interface?

Should we define an interface `IPolicyRegistry : IDictionary<string, Policy>`?  

_Initial view_: yes.  Users could provide their own alternative `IPolicyRegistry` implementations, eg backed by storage; stubbing out Polly for unit tests; taking a more specific view on what the `string` key represented; etc.

### What should the key represent?

_Suggest_: Polly's default implementation should be agnostic about what the string `key`s represent; allow users flexibility.

Some standard patterns might however emerge - for example, a `PolicyRegistry` where the keys correspond to the `ExecutionKey` (call site key), or `PolicyKey` (individual policy instance).  

## Views?

?

## Up-for-grabs

Marking this one 'up-for-grabs' (after community input on above topics ...).

**Comment by reisenberger** [comment]
An example of why `IPolicyRegistry` useful: it would allow for simple stubbing-out of Polly in unit tests.  

Consider an application architecture which defines policies in the registry on start-up; then passes the `PolicyRegistry` instance  by dependency injection to components needing a policy.  
 
With the new [`NoOpPolicy`](https://github.com/App-vNext/Polly/wiki/NoOp), this architecture would facilitate a very simple `NoOpPolicyRegistry : IPolicyRegistry` stub, to stub all policies out for unit tests on the application components:

```csharp
public class NoOpPolicyRegistry : IPolicyRegistry
{
    private static NoOpPolicy noop = Policy.NoOp();

    public Policy this[string key]
    {
        get { return noop; } // Intentionally override policies normally provided, with NoOp.
        set { /* do nothing */ }
    }

    // Other members ...
}
 
```

**Comment by reisenberger** [comment]
For future reference: Discussion [in our slack channel](https://pollytalk.slack.com/archives/general/p1487050660001004) moved this towards using a simpler (more reduced) `IPolicyRegistry<Key, TPolicy>` interface than the full `IDictionary<,>`.  (Further comment welcome)

Implementation in progress; PR expected.

**Comment by reisenberger** [comment]
Closed via #231 and #263
